src/Chainweb/Pact/PactService.hs

```diff
@@ -607,38 +620,49 @@ withParentBlockData phe action = action
     (BlockHeight !bh) = succ $ _blockHeight phe
     (BlockHash !ph) = _blockHash phe
 
+-- | Execute a block.
+--
+-- Precondition: restoreCheckpointer
+-- Postcondition: finalizeCheckpointer save
+--
 playOneBlock
     :: BlockHeader
     -> PayloadData
     -> PactDbEnv'
     -> PactServiceM cas PayloadWithOutputs
 playOneBlock currHeader plData pdbenv = do
-    -- precondition: restore has been called already
-    --
-    -- TODO: check precondition?
     miner <- decodeStrictOrThrow' (_minerData $ _payloadDataMiner plData)
     trans <- liftIO $ transactionsFromPayload plData
     !results <- withBlockData currHeader $ execTransactions pBlock miner trans pdbenv
-    finalizeCheckpointer (flip save bHash)   -- caller must restore again
+    finalizeCheckpointer (flip save bHash)
     psStateValidated .= Just currHeader
     return $! toPayloadWithOutputs miner results
-
   where
     bHash = _blockHash currHeader
     bParent = _blockParent currHeader
     isGenesisBlock = isGenesisBlockHeader currHeader
     pBlock = if isGenesisBlock then Nothing else Just bParent
 
+-- | @rewindTo mb act@ rewinds the pact state to @mb@ and runs @act@ on it.
+--
+-- If @mb@ is 'Nothing', it rewinds to the genesis block.
+--
+-- Postcondition: restoreCheckpointer
+--
 rewindTo
-    :: forall a cas . PayloadCas cas
+    :: forall cas . PayloadCas cas
     => Maybe (BlockHeight, ParentHash)
-    -> (PactDbEnv' -> PactServiceM cas a)
-    -> PactServiceM cas a
-rewindTo mb act = do
+        -- ^ The block height @height@ to which to restore and the parent header
+        -- @parentHeader@.
+        --
+        -- It holds that @(_blockHeight parentHeader == pred height)@
+
+    -> PactServiceM cas PactDbEnv'
+rewindTo mb = do
```

Comment:
Rewind used to take continuation. That continuation was wrapped in the exception handler bracket in `playFork`. But only handling lookup failures for last block, the continuation isn't in that bracket any more and `rewindTo` can become a "normal" function that returns a result directly.